File: CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Grid;


namespace WindowsApplication1
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
        private DevExpress.XtraGrid.GridControl gridControl1;
        private DevExpress.XtraGrid.Views.Grid.GridView gridView1;
        private System.ComponentModel.IContainer components;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.gridControl1 = new DevExpress.XtraGrid.GridControl();
            this.gridView1 = new DevExpress.XtraGrid.Views.Grid.GridView();
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).BeginInit();
            this.SuspendLayout();
            // 
            // gridControl1
            // 
            this.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill;
            this.gridControl1.MainView = this.gridView1;
            this.gridControl1.Name = "gridControl1";
            this.gridControl1.Size = new System.Drawing.Size(480, 270);
            this.gridControl1.TabIndex = 0;
            this.gridControl1.Text = "gridControl1";
            // 
            // gridView1
            // 
            this.gridView1.GridControl = this.gridControl1;
            this.gridView1.Name = "gridView1";
            this.gridView1.ShownEditor += new System.EventHandler(this.gridView1_ShownEditor);
            // 
            // Form1
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(480, 270);
            this.Controls.AddRange(new System.Windows.Forms.Control[] {
                                                                          this.gridControl1});
            this.Name = "Form1";
            this.Text = "Form1";
            this.Load += new System.EventHandler(this.Form1_Load);
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).EndInit();
            this.ResumeLayout(false);

        }
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

		private void Form1_Load(object sender, System.EventArgs e) {
            DataTable mTable = new DataTable();
            mTable.Columns.Add("ID", typeof(int));
            mTable.Columns.Add("check", typeof(bool));
            mTable.Columns.Add("data", typeof(string));

            DataTable dTable = new DataTable();
            dTable.Columns.Add("ID", typeof(int));
            dTable.Columns.Add("ChildID", typeof(int));
            dTable.Columns.Add("check", typeof(bool));
            dTable.Columns.Add("data", typeof(string));

            for(int i = 0; i < 10; i++)
                mTable.Rows.Add(new object[] {i, false, "Row " + i.ToString()});

            for(int i = 0; i < 100; i++)
                dTable.Rows.Add(new object[] {i, i /10, false, "Row " + i.ToString()});
            
            DataSet ds = new DataSet();
            ds.Tables.Add(mTable);
            ds.Tables.Add(dTable);
            ds.Relations.Add(new DataRelation("myRelation", mTable.Columns["ID"], dTable.Columns["ChildID"], true));
            gridControl1.DataSource = ds.Tables[0];
		}

        private void gridView1_ShownEditor(object sender, System.EventArgs e) {
            if((sender as GridView).ActiveEditor is CheckEdit) {
                CheckEdit edit = (sender as GridView).ActiveEditor as CheckEdit;
                edit.CheckedChanged += new System.EventHandler(myEdit_CheckedChanged);
            }
        }

        private void myEdit_CheckedChanged(object sender, System.EventArgs e) {
            UpdateDetailView(gridView1.FocusedRowHandle, (sender as CheckEdit).Checked);
        }

        private void UpdateDetailView(int rowHandle, bool state) {
            GridView dView = gridView1.GetDetailView(rowHandle, 0) as GridView;
            bool aCollapsed = dView == null;
            if (dView == null)
            {
                gridView1.ExpandMasterRow(rowHandle);
                dView = gridView1.GetDetailView(rowHandle, 0) as GridView;
            }
            if(dView != null)
                for(int i = 0; i < dView.DataRowCount; i++) 
                    dView.SetRowCellValue(i, dView.Columns["check"], state);
        }
	}
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Grid


Namespace WindowsApplication1
	''' <summary>
	''' Summary description for Form1.
	''' </summary>
	Public Class Form1
		Inherits System.Windows.Forms.Form
		Private gridControl1 As DevExpress.XtraGrid.GridControl
		Private WithEvents gridView1 As DevExpress.XtraGrid.Views.Grid.GridView
		Private components As System.ComponentModel.IContainer

		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()

			'
			' TODO: Add any constructor code after InitializeComponent call
			'
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Windows Form Designer generated code"
		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.gridControl1 = New DevExpress.XtraGrid.GridControl()
			Me.gridView1 = New DevExpress.XtraGrid.Views.Grid.GridView()
			CType(Me.gridControl1, System.ComponentModel.ISupportInitialize).BeginInit()
			CType(Me.gridView1, System.ComponentModel.ISupportInitialize).BeginInit()
			Me.SuspendLayout()
			' 
			' gridControl1
			' 
			Me.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill
			Me.gridControl1.MainView = Me.gridView1
			Me.gridControl1.Name = "gridControl1"
			Me.gridControl1.Size = New System.Drawing.Size(480, 270)
			Me.gridControl1.TabIndex = 0
			Me.gridControl1.Text = "gridControl1"
			' 
			' gridView1
			' 
			Me.gridView1.GridControl = Me.gridControl1
			Me.gridView1.Name = "gridView1"
'			Me.gridView1.ShownEditor += New System.EventHandler(Me.gridView1_ShownEditor);
			' 
			' Form1
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(480, 270)
			Me.Controls.AddRange(New System.Windows.Forms.Control() { Me.gridControl1})
			Me.Name = "Form1"
			Me.Text = "Form1"
'			Me.Load += New System.EventHandler(Me.Form1_Load);
			CType(Me.gridControl1, System.ComponentModel.ISupportInitialize).EndInit()
			CType(Me.gridView1, System.ComponentModel.ISupportInitialize).EndInit()
			Me.ResumeLayout(False)

		End Sub
		#End Region

		''' <summary>
		''' The main entry point for the application.
		''' </summary>
		<STAThread> _
		Shared Sub Main()
			Application.Run(New Form1())
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
			Dim mTable As New DataTable()
			mTable.Columns.Add("ID", GetType(Integer))
			mTable.Columns.Add("check", GetType(Boolean))
			mTable.Columns.Add("data", GetType(String))

			Dim dTable As New DataTable()
			dTable.Columns.Add("ID", GetType(Integer))
			dTable.Columns.Add("ChildID", GetType(Integer))
			dTable.Columns.Add("check", GetType(Boolean))
			dTable.Columns.Add("data", GetType(String))

			For i As Integer = 0 To 9
				mTable.Rows.Add(New Object() {i, False, "Row " & i.ToString()})
			Next i

			For i As Integer = 0 To 99
				dTable.Rows.Add(New Object() {i, i \10, False, "Row " & i.ToString()})
			Next i

			Dim ds As New DataSet()
			ds.Tables.Add(mTable)
			ds.Tables.Add(dTable)
			ds.Relations.Add(New DataRelation("myRelation", mTable.Columns("ID"), dTable.Columns("ChildID"), True))
			gridControl1.DataSource = ds.Tables(0)
		End Sub

		Private Sub gridView1_ShownEditor(ByVal sender As Object, ByVal e As System.EventArgs) Handles gridView1.ShownEditor
			If TypeOf (TryCast(sender, GridView)).ActiveEditor Is CheckEdit Then
				Dim edit As CheckEdit = TryCast((TryCast(sender, GridView)).ActiveEditor, CheckEdit)
				AddHandler edit.CheckedChanged, AddressOf myEdit_CheckedChanged
			End If
		End Sub

		Private Sub myEdit_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs)
			UpdateDetailView(gridView1.FocusedRowHandle, (TryCast(sender, CheckEdit)).Checked)
		End Sub

		Private Sub UpdateDetailView(ByVal rowHandle As Integer, ByVal state As Boolean)
			Dim dView As GridView = TryCast(gridView1.GetDetailView(rowHandle, 0), GridView)
			Dim aCollapsed As Boolean = dView Is Nothing
			If dView Is Nothing Then
				gridView1.ExpandMasterRow(rowHandle)
				dView = TryCast(gridView1.GetDetailView(rowHandle, 0), GridView)
			End If
			If dView IsNot Nothing Then
				For i As Integer = 0 To dView.DataRowCount - 1
					dView.SetRowCellValue(i, dView.Columns("check"), state)
				Next i
			End If
		End Sub
	End Class
End Namespace
